### PR TITLE
[Merged by Bors] - fix: remove lambda abstraction of `Nat.rec_zero` and `Nat.rec_add_one`

### PR DESCRIPTION
**Pull request by Komyyy**
```lean
theorem rec_zero {C : ℕ → Sort u} (h0 : C 0) (h : ∀ n, C n → C (n + 1)) :
    (Nat.rec h0 h : ∀ n, C n) 0 = h0 :=
  rfl
```
The above theorem is elaborated as follow:
```lean
theorem rec_zero {C : ℕ → Sort u} (h0 : C 0) (h : ∀ n, C n → C (n + 1)) :
    (fun n => Nat.rec h0 h n) 0 = h0 :=
  rfl
```
This form of the theorem isn't generic. This PR fixes this problem.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
Could you add a porting note, explaining why this change is necessary relative to mathlib3?

File: Mathlib/Data/Nat/Basic.lean [hunk 1]
```diff
@@ -373,16 +373,18 @@ This section is here due to dependencies -- the lemmas here require some of the
 proved above, and some of the results in later sections depend on the definitions in this section.
 -/
 
+-- Porting note: The type ascriptions of These two theorems are removed because
+--               lambda abstractions are newly inserted in Lean4.
```
Comment (semorrison) [inline Mathlib/Data/Nat/Basic.lean:377]: I'm confused. The lambdas are there in mathlib3, aren't they?

Comment (Komyyy) [reply Mathlib/Data/Nat/Basic.lean:377]: They are not there in mathlib3: https://leanprover-community.github.io/mathlib_docs/data/nat/basic.html#nat.rec_zero

Comment (semorrison) [reply Mathlib/Data/Nat/Basic.lean:377]: Haha, confusing: click through to "source" on the link you provided. The lambdas are really there! :-)

Comment (Komyyy) [reply Mathlib/Data/Nat/Basic.lean:377]: `(nat.rec h0 h : Π n, C n)` isn't elabolated to lambda in Lean3:
```lean
@[simp] lemma rec_zero {C : ℕ → Sort u} (h0 : C 0) (h : ∀ n, C n → C (n + 1)) :
  (nat.rec h0 h : Π n, C n) 0 = h0 :=
rfl

#print rec_zero
/-
@[_refl_lemma, _simp_cache, simp]
theorem nat.rec_zero : ∀ {C : ℕ → Sort u} (h0 : C 0) (h : Π (n : ℕ), C n → C (n + 1)), nat.rec h0 h 0 = h0 :=
λ {C : ℕ → Sort u} (h0 : C 0) (h : Π (n : ℕ), C n → C (n + 1)), rfl
-/
```
Here's no lambda.

Comment (semorrison) [reply Mathlib/Data/Nat/Basic.lean:377]: Thanks, sorry for my slowness here. :-)

Comment (semorrison) [reply Mathlib/Data/Nat/Basic.lean:377]:
```suggestion
-- Porting note: The type ascriptions of these two theorems need to be changed,
-- as mathport wrote a lambda that wasn't there in mathlib3, that prevents `simp` applying them.
```

**Comment by semorrison** [comment]
bors d+

**Pull request closed without merging**